### ISSUE TITLE
~~~ When Your Stack Traces Are Illegible ~~~

### ISSUE THREAD

**Issue by emars** [haml/haml#906]
```
/Users/emars/hyre/hyre-rails/app/views/organizers/partials/_request_staff_modal.html.haml:45: syntax error, unexpected ')' ));}\n", 0, false);end;_hamlou... ^ /Users/emars/hyre/hyre-rails/app/views/organizers/partials/_request_staff_modal.html.haml:45: unknown regexp options - dv /Users/emars/hyre/hyre-rails/app/views/organizers/partials/_request_staff_modal.html.haml:45: syntax error, unexpected $undefined ... </div>\n </div>\n <button class='p... ... ^ /Users/emars/hyre/hyre-rails/app/views/organizers/partials/_request_staff_modal.html.haml:45: syntax error, unexpected keyword_class, expecting keyword_do or '{' or '(' ...div>\n <button class='pull-right btn btn-success'... ... ^ /Users/emars/hyre/hyre-rails/app/views/organizers/partials/_request_staff_modal.html.haml:45: syntax error, unexpected $undefined ...='pull-right btn btn-success'>\n Invite\n ... ... ^ /Users/emars/hyre/hyre-rails/app/views/organizers/partials/_request_staff_modal.html.haml:45: unknown regexp options - dv unmatched close parenthesis: /button>\n", -2, false);end;end;end;_hamlout.push_text(" </i /Users/emars/hyre/hyre-rails/app/views/organizers/partials/_request_staff_modal.html.haml:45: syntax error, unexpected $undefined ...ut.push_text(" </div>\n </div>\n </div>... ... ^ /Users/emars/hyre/hyre-rails/app/views/organizers/partials/_request_staff_modal.html.haml:45: unknown regexp options - dv /Users/emars/hyre/hyre-rails/app/views/organizers/partials/_request_staff_modal.html.haml:45: syntax error, unexpected $undefined ...\n </div>\n </div>\n </div>\n </div>\n</div>... ... ^ /Users/emars/hyre/hyre-rails/app/views/organizers/partials/_request_staff_modal.html.haml:45: unknown regexp options - dv /Users/emars/hyre/hyre-rails/app/views/organizers/partials/_request_staff_modal.html.haml:45: syntax error, unexpected $undefined ... </div>\n </div>\n </div>\n</div>\n", -6, false);::Haml... ... ^ /Users/emars/hyre/hyre-rails/app/views/organizers/partials/_request_staff_modal.html.haml:45: unterminated regexp meets end of file /Users/emars/hyre/hyre-rails/app/views/organizers/partials/_request_staff_modal.html.haml:45: syntax error, unexpected end-of-input, expecting keyword_end
```

~~~~~ jUsT HaMl ThInGs ~~~~~

In all seriousness what I the best way to go about debugging HAML views that don't render?

I'm on a new project that has some large HAML files and I need to be able to maintain them but often when I make changes, I get a huge stack trace which doesn't help my find the source of the error.

I have tried a couple gems such as `haml-lint` but they don't catch the cause of errors like this.

**Comment by k0kubun** [comment]
In Haml template debugging, all you need to see is only the top of backtrace. i.e., the line `/Users/emars/hyre/hyre-rails/app/views/organizers/partials/_request_staff_modal.html.haml:45` is broken. You don't need to see others.

This backtrace means that your partial Ruby code in that line raises SyntaxError. And how it's wrong is written in the error message.

For `syntax error, unexpected ')'`, `unknown regexp options - dv`, `syntax error, unexpected $undefined`, `syntax error, unexpected keyword_class, expecting keyword_do or '{' or '('`, `syntax error, unexpected $undefined`, `unknown regexp options - dv unmatched close parenthesis`, `syntax error, unexpected ')'`, ...etc, they are not related to Haml. It's just "Ruby" syntax error in the line. The error message is very clear and should be easy to fix.

<hr>

But `syntax error, unexpected end-of-input, expecting keyword_end` is a little complicated. In some cases, Haml appends `end` keyword. Maybe it's not the case for your current template. You may want to see compiled code by `bundle exec haml -d -t ugly -e /Users/emars/hyre/hyre-rails/app/views/organizers/partials/_request_staff_modal.html.haml` and it may lack some `end` keyword in somewhere.

**Comment by emars** [comment]
Hey @k0kubun, thanks for responding so promptly. You were correct that that line in question was broken (45). The indentation level was off inside a block. 

I also must have been missing a `do` statement somewhere, it was resolved by refactoring the view into partials which were easier to look at and manage. My issue was that the file I was dealing with was several hundred lines and the indentation level was getting very deep which was hard to reason about.

I apologize if I came across as facetious.
